src/probability_theory/density.lean

```diff
@@ -0,0 +1,301 @@
+the distribution of a random variable, and are useful for calculating probabilities and
+finding moments (although the latter is better achieved with moment generating functions).
+
+This file also defines the continuous uniform distribution and proves some properties about
+random variables with this distribution.
+
+## Main definitions
+
+* `measure_theory.measure.has_pdf` : A random variable `X : α → E` is said to `has_pdf` with
+  respect to the measure `ℙ` on `α` and `μ` on `E` if there exists a measurable function `f`
+  such that the push-forward measure of `ℙ` along `X` equals `μ.with_density f`.
+* `measure_theory.measure.pdf` : If `X` is a random variable that `has_pdf X ℙ μ`, then `pdf X`
+  is the measurable function `f` such that the push-forward measure of `ℙ` along `X` equals
+  `μ.with_density f`.
+* `measure_theory.measure.pdf.uniform` : A random variable `X` is said to follow the uniform
+  distribution if it has a constant probability density function with a compact, non-null support.
+
+## Main results
+
+* `measure_theory.measure.pdf.integral_mul_eq_integral'` : Law of the unconscious statistician,
+  i.e. if a random variable `X : α → E` has pdf `f`, then `𝔼(g(X)) = ∫ x, g x * f x dx` for
+  all measurable `g : E → ℝ`.
+* `measure_theory.measure.pdf.integral_mul_eq_integral` : A real-valued random variable `X` with
+  pdf `f` has expectation `∫ x, x * f x dx`.
+* `measure_theory.measure.pdf.uniform.integral_eq` : If `X` follows the uniform distribution with
+  its pdf having support `s`, then `X` has expectation `(λ s)⁻¹ * ∫ x in s, x dx` where `λ`
+  is the Lebesgue measure.
+
+## TODOs
+
+Ultimately, we would also like to define characteristic functions to describe distributions as
+it exists for all random variables. However, to define this, we will need Fourier transforms
+which we currently do not have.
+-/
+
+noncomputable theory
+open_locale classical measure_theory nnreal ennreal
+
+namespace measure_theory
+
+open topological_space
+
+variables {α E : Type*} [normed_group E] [measurable_space E] [second_countable_topology E]
+  [normed_space ℝ E] [complete_space E] [borel_space E]
+
+namespace measure
+
+/-- A random variable `X : α → E` is said to `has_pdf` with respect to the measure `ℙ` on `α` and
+`μ` on `E` if there exists a measurable function `f` such that the push-forward measure of `ℙ`
+along `X` equals `μ.with_density f`. -/
+class has_pdf {m : measurable_space α} (X : α → E)
+  (ℙ : measure α) (μ : measure E . volume_tac) : Prop :=
+(pdf' : ∃ (f : E → ℝ≥0∞), measurable f ∧ map X ℙ = μ.with_density f)
+
+/-- If `X` is a random variable that `has_pdf X ℙ μ`, then `pdf X` is the measurable function `f`
+such that the push-forward measure of `ℙ` along `X` equals `μ.with_density f`. -/
+def pdf {m : measurable_space α} (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) :=
+if hX : has_pdf X ℙ μ then classical.some hX.pdf' else 0
```

Comment:
With the current `has_pdf` and `pdf` definitions, if `X` is not measurable then it `has_pdf` since `map X P µ` is 0, and its pdf is some arbitrary function which is almost zero everywhere. I think that we should set the pdf to 0 exactly and see two ways to do that:
1) change `has_pdf` to require that `X` is measurable.
or
2) change the condition in the `if` that defines `pdf` to require that `X` is measurable and `has_pdf`

Since having `has_pdf` is not useful for non-measurable functions, I propose solution 1. A consequence of this change would be that many `measurable X` hypotheses can be dropped in this file.

Here are a few lemmas I wrote while investigating this question, which might be useful:
```
lemma pdf_undef {m : measurable_space α} {ℙ : measure α} {μ : measure E} {X : α → E}
  (h : ¬ has_pdf X ℙ μ) :
  pdf X ℙ μ = 0 :=
by simp only [pdf, dif_neg h]

/-- TODO: does something like this exists? -/
lemma with_density_eq_zero {μ : measure E} {f : E → ℝ≥0∞}
  (hf : ae_measurable f μ) (h : μ.with_density f = 0) :
  f =ᵐ[μ] 0 :=
begin
  refine ennreal.ae_eq_zero_of_lintegral_rpow_eq_zero zero_lt_one hf _,
  simp_rw ennreal.rpow_one,
  rw [← set_lintegral_univ, ← with_density_apply _ measurable_set.univ, h],
  simp,
end

lemma pdf_ae_eq_zero_of_not_measurable {m : measurable_space α} {ℙ : measure α} {μ : measure E}
  {X : α → E} (hX : ¬ measurable X) :
  pdf X ℙ μ =ᵐ[μ] 0 :=
begin
  by_cases hX_pdf : has_pdf X ℙ μ,
  { haveI := hX_pdf,
    have h_eq := map_eq_with_density_pdf X ℙ μ,
    rw map_of_not_measurable hX at h_eq,
    exact with_density_eq_zero (measurable_pdf X ℙ μ).ae_measurable h_eq.symm, },
  { exact filter.eventually_of_forall (λ x, by rw pdf_undef hX_pdf), },
end

lemma measurable_of_pdf_ne_zero {m : measurable_space α} {ℙ : measure α} {μ : measure E} (X : α → E)
  (h : ¬ pdf X ℙ μ =ᵐ[μ] 0) : measurable X :=
by { by_contra hX, exact h (pdf_ae_eq_zero_of_not_measurable hX), }
```

Reply:
Is solution 1 changing the definition of `has_pdf` to 
```lean
class has_pdf {m : measurable_space α} {X : α → E} (hX : measurable X)
  (ℙ : measure α) (μ : measure E . volume_tac) : Prop :=
(pdf' : ∃ (f : E → ℝ≥0∞), measurable f ∧ map X ℙ = μ.with_density f)
```

Reply:
I'd rather change it to
```
class has_pdf {m : measurable_space α} (X : α → E)
  (ℙ : measure α) (μ : measure E . volume_tac) : Prop :=
(pdf' : measurable X ∧ ∃ (f : E → ℝ≥0∞), measurable f ∧ map X ℙ = μ.with_density f)
```

Reply:
Great! I made the change and was able to remove a lot of `measurable X` conditions